### PR TITLE
check_match: misc unifications and ICE fixes

### LINKED ISSUE #68393: ICE with associated constant as pattern in `let`
```rust
pub enum EFoo { A }

pub trait Foo {
    const X: EFoo;
}

struct Abc;

impl Foo for Abc {
    const X: EFoo = EFoo::A;
}

struct Def;
impl Foo for Def {
    const X: EFoo = EFoo::A;
}

pub fn test<A: Foo, B: Foo>(arg: EFoo) {
    let A::X = arg;
}

fn main() {
}
```

backtrace:
```

error: internal compiler error: encountered errors lowering pattern: [AssocConstInPattern(src/main.rs:19:9: 19:13)]
  --> src/main.rs:19:9
   |
19 |     let A::X = arg;
   |         ^^^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:356:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1057
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:476
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::real_drop_in_place
  15: core::ptr::real_drop_in_place
  16: core::ptr::real_drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: scoped_tls::ScopedKey<T>::set
  19: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (c0e02ad72 2020-01-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin
```

### LINKED ISSUE #68394: ICE with runtime value in pattern
```rust
fn main() {
    let x = 255u8;
    let 0u8..=x = 0;
}
```
backtrace:
```


error: internal compiler error: found bad range pattern `(Some((Constant { value: Const { ty: u8, val: Value(Scalar(0x00)) } }, None)), Some((Wild, None)))` outside of error recovery
 --> src/main.rs:3:9
  |
3 |     let 0u8..=x = 0;
  |         ^^^^^^^

error: internal compiler error: encountered errors lowering pattern: [NonConstPath(src/main.rs:3:15: 3:16)]
 --> src/main.rs:3:9
  |
3 |     let 0u8..=x = 0;
  |         ^^^^^^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:356:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1057
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:476
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::real_drop_in_place
  15: core::ptr::real_drop_in_place
  16: core::ptr::real_drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: scoped_tls::ScopedKey<T>::set
  19: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (c0e02ad72 2020-01-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground`.
```

### LINKED ISSUE #68396: ICE with `FloatBug` in pattern
```rust
fn main() {
    let 1234567890123456789012345678901234567890e-340: f64 = 0.0;
}
```
backtrace:
```

Standard Error

   Compiling playground v0.0.1 (/playground)
error: internal compiler error: encountered errors lowering pattern: [FloatBug]
 --> src/main.rs:4:9
  |
4 |     let 1234567890123456789012345678901234567890e-340: f64 = 0.0;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:356:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1057
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:476
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::real_drop_in_place
  15: core::ptr::real_drop_in_place
  16: core::ptr::real_drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: scoped_tls::ScopedKey<T>::set
  19: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (c0e02ad72 2020-01-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

### PR DESCRIPTION
**Pull request by Centril**
These are some unifications made as a by-product of working on `hir::ExprKind::Let`.

Fixes https://github.com/rust-lang/rust/issues/68396.
Fixes https://github.com/rust-lang/rust/issues/68394.
Fixes https://github.com/rust-lang/rust/issues/68393.

r? @oli-obk @matthewjasper

### CODE REVIEW

**Comment by oli-obk** [comment]
cc @rust-lang/wg-diagnostics

File: src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr [hunk 1]
```diff
@@ -0,0 +1,9 @@
+error[E0080]: runtime values cannot be referenced in patterns
```
Comment (estebank) [inline src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr]: I would like the wording of this error to be tweaked a bit.

File: src/test/ui/pattern/issue-68393-let-pat-assoc-constant.stderr [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  --> $DIR/issue-68393-let-pat-assoc-constant.rs:20:40
+   |
+LL | pub fn test<A: Foo, B: Foo>(arg: EFoo, A::X: EFoo) {
+   |                                        ^^^^
+
+error[E0158]: associated consts cannot be referenced in patterns
+  --> $DIR/issue-68393-let-pat-assoc-constant.rs:22:9
+   |
+LL |     let A::X = arg;
+   |         ^^^^
```
Comment (estebank) [inline src/test/ui/pattern/issue-68393-let-pat-assoc-constant.stderr]: Can we customize the "patterns" wording in fn args and let bindings? I would expect people that are unfamiliar with the language to get confused by this. At least a `note` with something like (but better) "`let` statements introduce new bindings to some pattern, normally a simple name which are also patterns".

File: src/test/ui/lint/lint-uppercase-variables.rs [hunk 3]
```diff
@@ -25,6 +25,16 @@ fn main() {
...
 
+    let Foo = foo::Foo::Foo;
+    //~^ ERROR variable `Foo` should have a snake case name
+    //~^^ WARN `Foo` is named the same as one of the variants of the type `foo::Foo`
+    //~^^^ WARN unused variable: `Foo`
+
+    fn in_param(Foo: foo::Foo) {}
+    //~^ ERROR variable `Foo` should have a snake case name
+    //~^^ WARN `Foo` is named the same as one of the variants of the type `foo::Foo`
+    //~^^^ WARN unused variable: `Foo`
```
Comment (estebank) [inline src/test/ui/lint/lint-uppercase-variables.rs]: Can we use `//~|` in this file?

**Review by estebank — APPROVED** [review]

File: src/test/ui/pattern/issue-68393-let-pat-assoc-constant.stderr (hunk 2 above)
Comment (Centril) [reply src/test/ui/pattern/issue-68393-let-pat-assoc-constant.stderr]: I think this is unlikely that a beginner would write this code personally. Using an associated constant in an irrefutable context seems fairly niche. In fact, I found this ICE by trying to unify the code, realizing that a check was missing in the irrefutable contexts. `A::X` also does not seem confusable for a binding. I would prefer not to preempt better diagnostics here until a beginner shows up.

If we are going to improve diagnostics, I'd rather add "defined here" `span_label`s for `X` and `A`.

File: src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr (hunk 1 above)
Comment (Centril) [reply src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr]: What's wrong with the current?

File: src/test/ui/lint/lint-uppercase-variables.rs (hunk 3 above)
Comment (Centril) [reply src/test/ui/lint/lint-uppercase-variables.rs]: I mostly copied existing code. Should we file an issue for follow-up?

Comment (estebank) [reply src/test/ui/lint/lint-uppercase-variables.rs]: Nah, its fine.

File: src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/pattern/issue-68394-let-pat-runtime-value.stderr]: I feel it is a bit jargony, but I guess it's no worse than other more common errors we emit.

**Pull request merged**